themes/qgrid/source/js/menu.js

```diff
@@ -15,41 +15,189 @@ function toggleVisibility(e) {
+	const p = document.getElementsByTagName('p');
+	let scrolled = false;
+	for (let searchTarget of [...h2, ...p]) {
+		const a = searchTarget.getElementsByTagName('a')[0];
+		if (a && a.textContent) {
+			searchTarget = a;
+		}
+		if (highlightText(searchTarget, search) && !scrolled) {
+			scrolled = true;
+			searchTarget.scrollIntoView();
+		}
+	}
+}
+
+function menuItemsSearch(search) {
 	const nav = document.getElementById('nav');
-	const li = nav.getElementsByTagName('li');
-
-	for (let i = 0; i < li.length; i++) {
-		const a = li[i].getElementsByTagName('a')[0];
-		if (a) {
-			if (a.text.toLowerCase().indexOf(filter) > -1) {
-				li[i].style.display = '';
-			} else {
-				li[i].style.display = 'none';
+	const titles = nav.querySelectorAll('.title');
+	for (let title of titles) {
+		const menuItem = title.parentElement;
+		for (let tag of getTags(menuItem)) {
+			if (search && tag.toLowerCase().indexOf(search) >= 0) {
+				appendTagText(menuItem, tag, search);
+				const dataShowTag = menuItem.getElementsByTagName('data-show-tag')[0].querySelector('a');
+				highlightText(dataShowTag, search);
+				break;
 			}
+			removeTagText(menuItem, search);
+		}
+		if (title.textContent[0] != '/') {
+			highlightText(title, search)
 		}
 	}
 }
 
+function appendTagText(menuItem, tag) {
+	menuItem.style.display = '';
+	if (menuItem.getElementsByTagName('data-show-tag').length === 0) {
+		const newTag = document.createElement('data-show-tag');
+		newTag.appendChild(document.createElement('a'));
+		menuItem.insertBefore(newTag, menuItem.children[0]);
+	}
+	const title = menuItem.querySelector('.title');
+	title.classList.add('menu-item');
+	if (title.textContent[0] != '/') {
+		title.prepend("/ ");
+	}
+	const dataShowTag = menuItem.getElementsByTagName('data-show-tag')[0];
+	dataShowTag.style.display = '';
+	dataShowTag.firstChild.textContent = formatTag(tag);
+	dataShowTag.firstChild.setAttribute('href', menuItem.querySelector('.title').getAttribute('href'));
+}
+
+function removeTagText(menuItem, search) {
+	const title = menuItem.querySelector('.title');
+	menuItem.style.display = (search) ? 'none' : '';
+	const dataShowTag = menuItem.getElementsByTagName('data-show-tag')[0];
+	if (dataShowTag) {
+		dataShowTag.style.display = 'none';
+	}
+	title.classList.remove('menu-item');
+	title.textContent = title.textContent.replace("/ ", '');
+}
+
+function highlightText(item, search) {
+	const { textContent } = item;
+	const searchContains = new RegExp(search, 'ig');
+	if (search && searchContains.test(textContent)) {
+		item.innerHTML = textContent.replace(searchContains, elem => `<span class="highlight">${elem}</span>`);
+		item.parentElement.style.display = '';
+		return true;
+	}
+	item.innerHTML = textContent;
+}
+
+function formatTag(tag) {
+	let tagsContent = getSplitText(tag);
+	const matchingText = matchText(tagsContent);
+	const index = tagsContent.indexOf(matchingText);
+	if (tagsContent.length > 3) {
```

Comment:
this is magic number - unreadable please make like, please make names for this numbers